rholang/examples/linking/packages/BalanceMap.rho

```diff
@@ -0,0 +1,75 @@
+export BalanceMap in {
+  import IMapApi in {
+    new 
+      IMapEmptyMap, IMapUpdated, IMapGetOrElse
+    in {
+      IMapApi("EmptyMap", IMapEmptyMap) | IMapApi("GetOrElse", IMapGetOrElse) |
+      IMapApi("Updated", IMapUpdated) | 
+      for(
+          emptyMap <- IMapEmptyMap; getOrElse <- IMapGetOrElse;
+          updated <- IMapUpdated
+      ){
+        
+        contract BalanceMap(return) = {
+          new query, sub, add, transfer, dataStore, api in {
+            emptyMap(dataStore) |
+            contract query(key, return) = {
+              for(map <- dataStore) {
+                dataStore(map) | getOrElse(key, map, 0, return)
+              }
+            } |
+            contract sub(key, amount, success) = {
+              match amount > 0 with true => {
+                new balanceCh in {
+                  for(map <- dataStore) {
+                    getOrElse(key, map, 0, balanceCh) | for(balance <- balanceCh) {
+                      match balance >= amount with true => {
+                        updated(map, key, balance - amount, dataStore) |
+                        success(true)
+                      }
+                      _ => { dataStore(map) | success(false) }
+                    }
+                  }
+                }
+              }
+              _ => { success(false) }
+            } |
+            contract add(key, amount, success) = {
+              match amount > 0 with true => {
+                new balanceCh in {
+                  for(map <- dataStore) {
+                    getOrElse(key, map, 0, balanceCh) | for(balance <- balanceCh) {
+                      updated(map, key, balance + amount, dataStore) |
+                      success(true)
+                    }
+                  }
+                }
+              }
+              _ => { success(false) }
+            } |
+            contract transfer(from, to, amount, success) = {
+              new result in{
+                sub(from, amount, result) | for(r <- result) {
+                  match r with true => {
+                    add(to, amount, success)
```

Comment:
Imagine the following attack that allows for negative balances.

The initial balances are as follows.

```
balance of me: 10
balance of they: 0
```

And the following call is executed.
 
`new rtn { transfer("me", "they", 10, rtn) | transfer("me", "they", 10, rtn)}`

I'm going to annotate any subcall of the first transfer with _1 and the second transfer with _2.

```
0. The first transfer call is executed - I'm going to annotate any subcall here with _1
1. success(true) executes in sub_1
2. updated(map, key, balance + amount, dataStore) executes in add_1
3. success(true) executes in add_1
4. success(true) executes in transfer_1
5. updated(map, key, balance - amount, dataStore) executes in sub_2 (note that the balance is still greater than 0 because the equivalent first call in sub_1 hasn't actually been executed yet)
6. success(true) executes in sub_2
7. updated(map, key, balance + amount, dataStore) executes in add_2
8. success(true) executes in transfer_2
9. updated(map, key, balance - amount, dataStore) executes in sub_1 (note the balance is now -10).
```
Note of course you can sequence the above code as follows, but we are assuming the attacker gets to write their own code.

`new rtn { transfer("me", "they", 10, rtn) | for (rtn2 <- rtn) { transfer("me", "they", 10, rtn2) } }`

Reply:
I think the solution is to have the `updated` call take in a return channel and block the success call until something comes back on the return channel.

Reply:
Btw we should keep a copy of this though. I want to see if we can catch this with behavioural types down the road.

Reply:
That execution order is not possible because nothing is in `dataStore` until after `updated(map, key, balance - amount, dataStore)` executes in sub_1. Therefore, every other call to the map would be blocked until `dataStore` is refilled with the updated balances.

`updated` does have a return channel, it is the one where the new map is sent (the last argument to the contract). So I could rewrite things to be 
```
new lock in {
  updated(map, key, balance - amount, lock) | for(newMap <- lock) {
    dataStore(newMap) | success(true)
  }
}
```
But that makes no difference because all other `BalanceMap` related calls will be blocked until `dataStore` is refilled anyways, which in the current version requires the `updated` call to have completed.

Reply:
I see. My eyes saw

`updated(...) | success(true)`

with only a guess to what updated actually does and I immediately thought that "oh this must be parallel code". I'm probably guilty of this in token.rho, but I think we need to make it more obvious that we are writing sequential code when we want things to be sequential. For example,

```
dataStore!(map) |
updated(map, key, balance - amount, lock) |
for (_ <- lock) {
     success(true)
}
```

I'm assuming here that map is just a pointer and changing the balance in parallel with storing back the map works.

Reply:
^ In a follow up discussion with Mike Birch, this isn't quite what I want either. The consume/replace pattern remains.